### ISSUE TITLE
tf.broadcast_to abort() and core dump when given shape value overflows int32

### ISSUE THREAD

**Issue by leeyeetonn** [tensorflow/tensorflow#40138]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:NA
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0 & v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: 3.7.6
- Bazel version (if compiling from source):NA
- GCC/Compiler version (if compiling from source):NA
- CUDA/cuDNN version:NA
- GPU model and memory:NA

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
`tf.broadcast_to` aborts and core dump if given `shape` could overflow `int32`. 
The `abort()` only seems to happen when `input` is in `uint32` or `uint64`.
If `input` is other dtype like `int32` or `float32`, tensorflow would not abort, but instead throw an exception to report the issue.
**Describe the expected behavior**
Tensorflow should not abort() and core dump.
Tensorflow should give a more meaningful message when overflow occurs and make the error handling behavior consistent.
**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

With `input` being `uint32` or `uint64` and a `shape` big enough to overflow `int32`:
```python
import tensorflow as tf
import numpy as np

in_tensor = np.array([1, 2]).astype('uint32')  # would later cause abort() and core dump
shape = np.array([1e18, 2]).astype('int64') # make shape big enough to overflow int32
tf.broadcast_to(in_tensor, shape)
```
would cause the following:
> ...
2020-06-04 06:16:17.434615: F tensorflow/core/framework/tensor_shape.cc:345] Check failed: size >= 0 (-1486618624 vs. 0)
Aborted (core dumped)

When `input` is not `uint32` nor `uint64` and a `shape` big enough to overflow `int32`:
```python
import tensorflow as tf
import numpy as np

in_tensor = np.array([1, 2]).astype('float32')  # would later cause exception
shape = np.array([1e18, 2]).astype('int64') # make shape big enough to overflow int32
tf.broadcast_to(in_tensor, shape)
```
would cause exception thrown instead of abort() and core dump:
>...
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimension -1486618624 must be >= 0 [Op:BroadcastTo]

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.1.0, 2.2.0 and the session is going to crash. Is this the expected behavior?.Thanks!

**Comment by leeyeetonn** [comment]
Yes, the first case where tensorflow abort() and core dump will make colab session to crash.

**Comment by yongtang** [comment]
Added a PR #40294 for the fix.

**Comment by jvishnuvardhan** [comment]
@leeyeetonn When I ran your codes with `tf-nightly`, it throws a different error than you are expecting but it is not crashing as you noticed earlier. [Here](https://colab.research.google.com/gist/jvishnuvardhan/bf06e470a5c4ba8be3c13d142700b78e/untitled86.ipynb) is a gist for our reference. Thanks!

Please verify once and close the issue if this was resolved for you. thanks!

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on CPU up to tensorflow 2.3.4 (v2.3.3-137-gea90cf44f73).

> On GPU, it throws an "CUDA runtime implicit initialization on GPU:0 failed." Exception before tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb), and it throws an "OOM" Exception in tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) and later versions.

- `(tf.broadcast_to)`, `tf.compat.v1.broadcast_to`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.broadcast_to</code> in older versions</summary>

```python
import numpy as np
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


try:
    in_tensor = np.array([1, 2]).astype('uint32')  # would later cause abort() and core dump
    shape = np.array([1e18, 2]).astype('int64') # make shape big enough to overflow int32
    tf.compat.v1.broadcast_to(in_tensor, shape)
except Exception as e:
    print("Failed! Error:", str(e), flush=True)
else:
    print("Success!", flush=True)
```

On my CPU machine, the process aborts with a Check failed error, which is not expected.

```text
v2.3.3-137-gea90cf44f73 2.3.4
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:XLA_CPU:0', device_type='XLA_CPU'), PhysicalDevice(name='/physical_device:XLA_GPU:0', device_type='XLA_GPU')]
2023-09-09 11:14:44.879197: F tensorflow/core/framework/tensor_shape.cc:173] Check failed: val >= 0 (-1486618624 vs. 0)
Aborted (core dumped)
```

While on GPU, it throws a "CUDA runtime implicit initialization on GPU:0 failed." Exception without aborting the process.

```text
v2.3.3-137-gea90cf44f73 2.3.4
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:XLA_CPU:0', device_type='XLA_CPU'), PhysicalDevice(name='/physical_device:XLA_GPU:0', device_type='XLA_GPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Failed! Error: CUDA runtime implicit initialization on GPU:0 failed. Status: device kernel image is invalid
```
</details>

It seems to be fixed in tensorflow 2.4.0 (v2.4.0-rc4-71-g582c8d236cb) and later versions.
